### DISCUSSION TITLE
Adding support for serializing datetime.datetime instances

### DISCUSSION THREAD

**Discussion by jreyesr** [writer/writer-framework#40] [Ideas]
Hello! I've been exploring Streamsync (awesome tool, BTW!) and I found out that apparently `datetime.datetime` instances cannot be serialized and sent to the frontend.

<details>
  <summary>Minimal example</summary>

  ```py
  import datetime
  import streamsync as ss

  initial_state = ss.init_state({
    "now": datetime.datetime.now(),
  })
  ```
</details>

This causes the following error to appear:

<details>
  <summary>Error</summary>

```
Serialisation error

Traceback (most recent call last):
  File "/home/reyes/.local/lib/python3.10/site-packages/streamsync/core.py", line 239, in to_dict
    serialised_value = state_serialiser.serialise(value)
  File "/home/reyes/.local/lib/python3.10/site-packages/streamsync/core.py", line 129, in serialise
    raise StateSerialiserException(
streamsync.core.StateSerialiserException: Object of type <class 'datetime.datetime'> (MRO: ['datetime.datetime', 'datetime.date', 'builtins.object']) cannot be serialised.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/reyes/.local/lib/python3.10/site-packages/streamsync/app_runner.py", line 120, in _handle_session_init
    user_state = session.session_state.user_state.to_dict()
  File "/home/reyes/.local/lib/python3.10/site-packages/streamsync/core.py", line 241, in to_dict
    raise ValueError(
ValueError: Couldn't serialise value of type "<class 'datetime.datetime'>" for key "now".
```

![image](https://github.com/streamsync-cloud/streamsync/assets/23390438/00e35461-4b3e-4e84-9a49-e0e58910b807)
</details>

As far as I can tell, this is due to the [serialization code](https://github.com/streamsync-cloud/streamsync/blob/master/src/streamsync/core.py#L89) not supporting `datetime.datetime` instances

Background: I ran across this when trying to replicate [Streamlit's starter tutorial](https://docs.streamlit.io/library/get-started/create-an-app) in Streamsync. This tutorial has a Pandas DataFrame with a datetime column (it's a special Pandas datetime class, but it eventually inherits from Python's `datetime.datetime`)

Adding support for this would let users display data that has timestamp-style columns, such as "created on", "event time" and such.

Is there any interest in letting users pass datetime instances to the FE? I don't see any issues or discussions mentioning it, but it may be because Streamsync is fairly new. Any ideas?

**Comment by ramedina86** [comment]
Hi, I'm glad you like Streamsync!

As you mentioned, `datetime.datetime` cannot be serialised because there's no code that can handle it.

This would be easy to implement and definitely useful, so I'll open an issue.

Thanks for suggesting this.

**Comment by jreyesr** [comment]
For cleanup, this was resolved in https://github.com/streamsync-cloud/streamsync/issues/41
